packages/loopring_v3.js/src/protocol_v3.ts

```diff
@@ -0,0 +1,232 @@
+import BN from "bn.js";
+const fs = require("fs");
+import Web3 from "web3";
+
+interface StakeAmounts {
+  exchangeStake: BN;
+  protocolFeeStake: BN;
+}
+
+/**
+ * Processes all data of the Loopring v3 protocol contract.
+ */
+export class ProtocolV3 {
+  private web3: Web3;
+
+  private loopringAbi: string;
+
+  private loopringAddress: string;
+  private loopring: any;
+  private owner: string;
+
+  private implementationManagerAddress: string;
+  private version: string;
+  private enabled: boolean;
+
+  private syncedToEthereumBlockIdx: number;
+
+  private stakeAmounts: { [key: number]: StakeAmounts } = {};
+
+  /**
+   * Initializes a Protocol
+   * @param   web3                           The web3 instance that will be used to get the necessary data from Ethereum
+   * @param   implementationManagerAddress   The address of the implementation manager
+   * @param   loopringAddress                The address of the protocol address
+   * @param   version                        The version of the protocol
+   */
+  public async initialize(
+    web3: Web3,
+    loopringAddress: string,
+    implementationManagerAddress: string,
+    version: string
+  ) {
+    this.web3 = web3;
+    this.loopringAddress = loopringAddress;
+    this.implementationManagerAddress = implementationManagerAddress;
+    this.version = version;
+    this.enabled = true;
+
+    this.syncedToEthereumBlockIdx = 0;
+
+    const ABIPath = "ABI/version30/";
+    this.loopringAbi = fs.readFileSync(ABIPath + "ILoopringV3.abi", "ascii");
+
+    this.loopring = new web3.eth.Contract(JSON.parse(this.loopringAbi));
+    this.loopring.options.address = this.loopringAddress.toLowerCase();
+
+    // Get the latest owner
+    this.owner = await this.loopring.methods.owner().call();
+  }
+
+  /**
+   * Syncs the protocol up to (and including) the given Ethereum block index.
+   * @param   ethereumBlockTo   The Ethereum block index to sync to
+   */
+  public async sync(ethereumBlockTo: number) {
+    if (ethereumBlockTo <= this.syncedToEthereumBlockIdx) {
+      return;
+    }
+
+    // Process the events
+    const events = await this.loopring.getPastEvents("allEvents", {
+      fromBlock: this.syncedToEthereumBlockIdx + 1,
+      toBlock: ethereumBlockTo
```

Comment:
if the number of blocks is too many, will this cause an out-of-memory error? Are we supposed to sync to only a couple of blocks per call? if so, we probably want to check `toBlock - fromBlock <= MAX_BLOCK_PER_CALL`

Reply:
I think it can go out of memory, but a maximum can be hard to decide on the library level because the amount of memory that can be used is by the library is hard to predict. . I think the code using the library is better at deciding this and can do small ranges to play safe and increase it if memory use is sufficiently low.